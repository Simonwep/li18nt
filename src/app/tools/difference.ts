import {JSONArray, JSONObject, JSONValue, PropertyPath} from '@types';
import {keysFrom} from '@utils/keysFrom';
import {typeOfJsonValue} from '@utils/typeOfJsonValue';

export type Difference = {
    missing: PropertyPath[];
    conflicts: PropertyPath[];
};

/**
 * Pushes the item if not already present.
 * @param arr
 * @param el
 */
const pushUnique = (arr: PropertyPath[], el: PropertyPath): boolean => {
    outer: for (const item of arr) {
        if (item.length === el.length) {

            for (let i = 0; i < arr.length; i++) {
                if (item[i] !== el[i]) {
                    continue outer;
                }
            }

            return false;
        }
    }

    arr.push(el);
    return true;
};

/**
 * Compares an object to others
 * @param target
 * @param others
 */
const compare = (target: JSONObject, others: JSONObject[]): Difference => {
    const diff: Difference = {
        conflicts: [],
        missing: []
    };

    function handle<K extends string | number, T extends Record<K, JSONValue>, O extends Record<K, JSONArray>[]>(
        key: K,
        target: T,
        others: O,
        parent: PropertyPath = []
    ): void {
        const targetValue = target[key];
        const targetType = typeOfJsonValue(targetValue);

        // Property missing?
        if (targetType === 'undefined') {
            pushUnique(diff.missing, [...parent, key]);
            return;
        }

        // Compare with others
        for (const obj of others) {
            const objValue = obj[key];
            const objType = typeOfJsonValue(objValue);

            // Property missing, skip
            if (objType === 'undefined') {
                continue;
            }

            // Property-type mismatch?
            if (objType !== targetType) {
                pushUnique(diff.conflicts, [...parent, key]);
                continue;
            }

            // Child object?
            if (objType === 'object' && objType === targetType) {
                resolve(targetValue as JSONObject, [objValue] as Record<number, JSONArray>[], [...parent, key]);
                continue;
            }

            // Child array?
            if (objType === 'array') {

                // Length mismatch
                if ((targetValue as JSONArray).length !== (targetValue as JSONArray).length) {
                    pushUnique(diff.conflicts, [...parent, key]);
                    continue;
                }

                // Resolve
                resolve(targetValue as JSONArray, objValue as JSONArray, [...parent, key]);
            }
        }
    }

    function resolve<T extends JSONObject | JSONArray, O extends (T extends JSONObject ? JSONObject[] : JSONArray)>(
        target: T,
        others: O,
        parent: PropertyPath = []
    ): void {
        if (Array.isArray(target) && Array.isArray(others)) {
            const maxLength = Math.max(
                target.length,
                others.length
            );

            for (let i = 0; i < maxLength; i++) {
                handle(i, target as JSONArray, others as Record<number, JSONArray>[], parent);
            }
        } else {
            for (const key of keysFrom([target as JSONObject, ...(others as JSONObject[])])) {
                handle(key, target as JSONObject, others as Record<string, JSONArray>[], parent);
            }
        }
    }

    resolve(target, others);
    return diff;
};

/**
 * Finds the difference between given objects
 * @param objects
 */
export const difference = (objects: JSONObject[]): Difference[] => {

    // Create result objects
    const differences: Difference[] = [];
    for (let i = 0; i < objects.length; i++) {
        const target = objects[i];
        const others = [...objects];
        others.splice(i, 1);
        differences.push(compare(target, others));
    }

    return differences;
};
