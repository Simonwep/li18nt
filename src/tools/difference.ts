import {keysFrom} from '@utils/keysFrom';
import {typeOfJsonValue} from '@utils/typeOfJsonValue';

type JSONRecord = {[key: string]: JSONValue;};
type JSONValue = string | number | JSONRecord | JSONArray;
type JSONArray = string[] | number[] | JSONValue[];

export type PropertyPath = (string | number)[];
export type Difference = {
    missing: PropertyPath[];
    conflicts: PropertyPath[];
};

/**
 * Compares an object to others
 * @param target
 * @param others
 */
const compare = (target: JSONRecord, others: JSONRecord[]): Difference => {
    const diff: Difference = {
        conflicts: [],
        missing: []
    };

    function handle<K extends string | number, T extends Record<K, JSONValue>, O extends Record<K, JSONArray>[]>(
        key: K,
        target: T,
        others: O,
        parent: PropertyPath = []
    ): void {
        const targetValue = target[key];
        const targetType = typeOfJsonValue(targetValue);

        // Property missing?
        if (targetType === 'undefined') {
            diff.missing.push([...parent, key]);
            return;
        }

        // Compare with others
        for (const obj of others) {
            const objValue = obj[key];
            const objType = typeOfJsonValue(objValue);

            // Property missing, skip
            if (objType === 'undefined') {
                return;
            }

            // Property-type mismatch?
            if (objType !== targetType) {
                diff.conflicts.push([...parent, key]);
                return;
            }

            // Child object?
            if (objType === 'object' && objType === targetType) {
                resolve(targetValue as JSONRecord, [objValue] as Record<number, JSONArray>[], [...parent, key]);
                return;
            }

            // Child array?
            if (objType === 'array') {

                // Length mismatch
                if ((targetValue as JSONArray).length !== (targetValue as JSONArray).length) {
                    diff.conflicts.push([...parent, key]);
                    return;
                }

                // Resolve
                resolve(targetValue as JSONArray, objValue as JSONArray, [...parent, key]);
            }
        }
    }

    function resolve<T extends JSONRecord | JSONArray, O extends(T extends JSONRecord ? JSONRecord[] : JSONArray)>(
        target: T,
        others: O,
        parent: PropertyPath = []
    ): void {
        if (Array.isArray(target) && Array.isArray(others)) {
            const maxLength = Math.max(
                target.length,
                others.length
            );

            for (let i = 0; i < maxLength; i++) {
                handle(i, target as JSONArray, others as Record<number, JSONArray>[], parent);
            }
        } else {
            for (const key of keysFrom([target as JSONRecord, ...(others as JSONRecord[])])) {
                handle(key, target as JSONRecord, others as Record<string, JSONArray>[], parent);
            }
        }
    }

    resolve(target, others);
    return diff;
};

/**
 * Finds the difference between given objects
 * @param objects
 */
export const difference = (objects: JSONRecord[]): Difference[] => {

    // Create result objects
    const differences: Difference[] = [];
    for (let i = 0; i < objects.length; i++) {
        const target = objects[i];
        const others = [...objects];
        others.splice(i, 1);
        differences.push(compare(target, others));
    }

    return differences;
};
