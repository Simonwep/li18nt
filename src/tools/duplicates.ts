import {JSONArray, JSONObject, PropertyPath} from '@types';
import {startsWithDeep} from '@utils/containsDeep';
import {propertyPath} from '@utils/propertyPath';
import {typeOfJsonValue} from '@utils/typeOfJsonValue';

export interface DuplicatesConfig {
    ignore?: (PropertyPath | string)[];
}

export type Duplicates = Map<string, PropertyPath[]>;

/**
 * Finds duplicate keys in the given object.
 * @param object
 * @param conf Optional configuration
 */
export const duplicates = (object: JSONObject, conf?: DuplicatesConfig): Duplicates => {
    const duplicates: Map<string, PropertyPath[]> = new Map<string, PropertyPath[]>();
    const keys: string[] = [];

    const ignored = conf?.ignore?.map(v => {
        return Array.isArray(v) ? v : propertyPath(v);
    }) || [];

    const walk = (entry: JSONObject | JSONArray, parentPath: PropertyPath): void => {
        if (Array.isArray(entry)) {
            for (let i = 0; i < entry.length; i++) {
                const value = entry[i];

                if (typeOfJsonValue(value) === 'object') {
                    walk(value as JSONObject, [...parentPath, i]);
                }
            }
        } else {
            for (const key in entry) {
                if (Object.prototype.hasOwnProperty.call(entry, key)) {
                    const newKey = [...parentPath, key];
                    const value = entry[key];
                    const type = typeOfJsonValue(value);

                    if (type === 'object' || type === 'array') {

                        // Make it possible to ignore entire sub-trees
                        if (!startsWithDeep(ignored, newKey)) {
                            walk(value as JSONObject, newKey);
                        }
                    } else if (keys.includes(key)) {
                        const list = duplicates.get(key) || [];

                        // Check against ignored list
                        if (!startsWithDeep(ignored, newKey)) {
                            duplicates.set(key, [...list, newKey]);
                        }
                    } else {
                        keys.push(key);
                    }
                }
            }
        }
    };

    walk(object, []);
    return duplicates;
};
