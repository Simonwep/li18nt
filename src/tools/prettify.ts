import {JSONArray, JSONObject, JSONValue} from '@types';
import {typeOfJsonValue} from '@utils/typeOfJsonValue';

export type Indentation = 'tab' | number;

export interface PrettifyOptions {
    indent?: Indentation;
}

/**
 * Returns the sorted keys of an object as string array
 * @param obj
 */
const sortedKeys = (obj: JSONObject): string[] => {
    return Object.keys(obj).sort(
        (a, b) => a === b ? 0 : a.localeCompare(b)
    );
};

/**
 * Sorts an object by its keys, returns a string as ordering properties manually
 * is slow and we can't rely on their order after inserting them.
 * @param obj
 * @param space
 */
export const prettify = (obj: JSONObject, {indent = 4}: PrettifyOptions): string => {
    const spacer = indent === 'tab' ? '\t' : ' '.repeat(indent);
    let str = '{\n';

    const stringify = (v: JSONValue, indent: string): string | boolean | number | null => {
        const type = typeOfJsonValue(v);
        const nextIndent = indent + spacer;

        switch (type) {
            case 'object': {
                let str = '{\n';

                for (const key of sortedKeys(v as JSONObject)) {
                    str += `${nextIndent}"${key}": ${stringify((v as JSONObject)[key], nextIndent)},\n`;
                }

                return str.length > 2 ? `${str.slice(0, str.length - 2)}\n${indent}}` : '{}';
            }
            case 'array': {
                let str = '[\n';

                for (let i = 0; i < (v as JSONArray).length; i++) {
                    str += `${nextIndent + stringify((v as JSONArray)[i], nextIndent)},\n`;
                }

                return str.length > 2 ? `${str.slice(0, str.length - 2)}\n${indent}]` : '[]';
            }
            case 'boolean':
            case 'number':
            case 'null':
                return v as (boolean | number | null);
            case 'string':
                return JSON.stringify(v);
        }

        return null;
    };

    for (const key of sortedKeys(obj)) {
        str += `${spacer}"${key}": ${stringify(obj[key], spacer)},\n`;
    }

    return str.length > 2 ? `${str.slice(0, str.length - 2)}\n}` : '{}';
};

