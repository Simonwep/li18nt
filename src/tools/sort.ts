import {typeOfJsonValue} from '@utils/typeOfJsonValue';
import {JSONArray, JSONObject, JSONValue} from '../types';

/**
 * Returns the sorted keys of an object as string array
 * @param obj
 */
const sortedKeys = (obj: JSONObject): string[] => {
    return Object.keys(obj).sort(
        (a, b) => a === b ? 0 : a > b ? 1 : -1
    );
};

/**
 * Sorts an object by its keys, returns a string as ordering properties manually
 * is slow and we can't rely on their order after inserting them.
 * @param obj
 * @param space
 */
export const sort = (obj: JSONObject, space?: number | string): string => {
    const spacer = typeof space === 'number' ? ' '.repeat(space) : typeof space === 'string' ? space : '';
    let str = '{\n';

    const stringify = (v: JSONValue, indent: string): string | boolean | number | null => {
        const type = typeOfJsonValue(v);
        const nextIndent = indent + spacer;

        switch (type) {
            case 'object': {
                let str = '{\n';

                for (const key of sortedKeys(v as JSONObject)) {
                    str += `${nextIndent}"${key}": ${stringify((v as JSONObject)[key], nextIndent)},\n`;
                }

                return str.length > 2 ? `${str.slice(0, str.length - 2)}\n${indent}}` : '{}';
            }
            case 'array': {
                let str = '[\n';

                for (let i = 0; i < (v as JSONArray).length; i++) {
                    str += `${nextIndent + stringify((v as JSONArray)[i], nextIndent)},\n`;
                }

                return str.length > 2 ? `${str.slice(0, str.length - 2)}\n${indent}]` : '[]';
            }
            case 'boolean':
            case 'number':
            case 'null':
                return v as (string | boolean | number);
            case 'string':
                return `"${v}"`;
        }

        return null;
    };

    for (const key of sortedKeys(obj)) {
        str += `${spacer}"${key}": ${stringify(obj[key], spacer)},\n`;
    }

    return str.length > 2 ? `${str.slice(0, str.length - 2)}\n}` : '{}';
};

