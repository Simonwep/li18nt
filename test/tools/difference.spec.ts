import {difference} from '../../src/tools/difference';

describe('[tool] difference', () => {

    it('Should find the difference between three, two-dimensional object', () => {
        expect(difference([
            {a: 10, b: '20'},
            {c: 10, b: 20},
            {c: 10, z: 200}
        ])).toMatchSnapshot();
    });

    it('Should work with nested objects', () => {
        expect(difference([
            {
                a: 10,
                b: {u: 'u'},
                x: {
                    p: {v: 10}
                }
            },
            {
                c: 10,
                b: 20,
                x: {
                    p: {v: 10, k: 100},
                    i: 10
                }
            },
            {
                c: '10',
                b: 20,
                x: {
                    p: {v: 10, k: 'hello', z: 20},
                    i: 10
                }
            }
        ])).toMatchSnapshot();
    });

    it('Should work with nested objects and arrays', () => {
        expect(difference([
            {
                a: 10,
                b: {u: 'u'},
                x: [{l: 10}, 'hello']
            },
            {
                a: 10,
                b: {u: 'u'},
                x: [{l: 10}, 5],
                l: 20
            },
            {
                a: 10,
                b: {u: 'u'},
                x: [{l: 10}, 'hello', {'5': 100}]
            }
        ])).toMatchSnapshot();
    });
});
