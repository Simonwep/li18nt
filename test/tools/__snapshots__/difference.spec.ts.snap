// Vitest Snapshot v1

exports[`Test difference tool > Should find the difference between three, two-dimensional object 1`] = `
[
  {
    "conflicts": [
      [
        "b",
      ],
    ],
    "missing": [
      [
        "c",
      ],
      [
        "z",
      ],
    ],
  },
  {
    "conflicts": [
      [
        "b",
      ],
    ],
    "missing": [
      [
        "a",
      ],
      [
        "z",
      ],
    ],
  },
  {
    "conflicts": [],
    "missing": [
      [
        "a",
      ],
      [
        "b",
      ],
    ],
  },
]
`;

exports[`Test difference tool > Should work with nested objects 1`] = `
[
  {
    "conflicts": [
      [
        "b",
      ],
    ],
    "missing": [
      [
        "x",
        "p",
        "k",
      ],
      [
        "x",
        "i",
      ],
      [
        "c",
      ],
    ],
  },
  {
    "conflicts": [
      [
        "c",
      ],
      [
        "b",
      ],
      [
        "x",
        "p",
        "k",
      ],
    ],
    "missing": [
      [
        "x",
        "p",
        "z",
      ],
      [
        "a",
      ],
    ],
  },
  {
    "conflicts": [
      [
        "c",
      ],
      [
        "b",
      ],
      [
        "x",
        "p",
        "k",
      ],
    ],
    "missing": [
      [
        "a",
      ],
    ],
  },
]
`;

exports[`Test difference tool > Should work with nested objects and arrays 1`] = `
[
  {
    "conflicts": [
      [
        "x",
        0,
      ],
    ],
    "missing": [
      [
        "x",
        2,
      ],
      [
        "l",
      ],
    ],
  },
  {
    "conflicts": [
      [
        "x",
        0,
      ],
    ],
    "missing": [
      [
        "x",
        2,
      ],
    ],
  },
  {
    "conflicts": [
      [
        "x",
        0,
      ],
    ],
    "missing": [
      [
        "l",
      ],
    ],
  },
]
`;
